# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [1]:
import pandas as pd
import numpy as np

In [4]:
# Input the data set 

url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv'
df = pd.read_csv(url)

In [5]:
df

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.

In [10]:
df.shape
print(f"Number of rows: {df.shape[0]}")
print(f"Number of columns: {df.shape[1]}")

Number of rows: 4008
Number of columns: 11


- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              1071 non-null   object 
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 344.6+ KB


In [13]:
# types of data
df.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

- Customer -> Object (Correct)
- ST       -> Object (Correct)
- Gender   -> Object (Correct)
-Education                     object
- Customer Lifetime Value       object (Not correct, but it's for the NaN) In my opinion, this should be type float for next calculations 
- Income                       float64 (Correct)
- Monthly Premium Auto         float64 (Correct)
- Number of Open Complaints     object 
- Policy Type                   object (Correct)
- Vehicle Class                 object (Correct)
- Total Claim Amount           float64 (Correct)

In [17]:
df["Number of Open Complaints"].unique()

array(['1/0/00', '1/2/00', '1/1/00', '1/3/00', '1/5/00', '1/4/00', nan],
      dtype=object)

In [14]:
df.info

<bound method DataFrame.info of      Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 7

In [15]:
df

Customer          ST GENDER             Education  \
0     RB50392  Washington    NaN                Master   
1     QZ44356     Arizona      F              Bachelor   
2     AI49188      Nevada      F              Bachelor   
3     WW63253  California      M              Bachelor   
4     GA49547  Washington      M  High School or Below   
...       ...         ...    ...                   ...   
4003      NaN         NaN    NaN                   NaN   
4004      NaN         NaN    NaN                   NaN   
4005      NaN         NaN    NaN                   NaN   
4006      NaN         NaN    NaN                   NaN   
4007      NaN         NaN    NaN                   NaN   

     Customer Lifetime Value   Income  Monthly Premium Auto  \
0                        NaN      0.0                1000.0   
1                 697953.59%      0.0                  94.0   
2                1288743.17%  48767.0                 108.0   
3                 764586.18%      0.0                 106.0   
4                 536307.65%  36357.0                  68.0   
...                      ...      ...                   ...   
4003                     NaN      NaN                   NaN   
4004                     NaN      NaN                   NaN   
4005                     NaN      NaN                   NaN   
4006                     NaN      NaN                   NaN   
4007                     NaN      NaN                   NaN   

     Number of Open Complaints     Policy Type  Vehicle Class  \
0                       1/0/00   Personal Auto  Four-Door Car   
1                       1/0/00   Personal Auto  Four-Door Car   
2                       1/0/00   Personal Auto   Two-Door Car   
3                       1/0/00  Corporate Auto            SUV   
4                       1/0/00   Personal Auto  Four-Door Car   
...                        ...             ...            ...   
4003                       NaN             NaN            NaN   
4004                       NaN             NaN            NaN   
4005                       NaN             NaN            NaN   
4006                       NaN             NaN            NaN   
4007                       NaN             NaN            NaN   

      Total Claim Amount  
0               2.704934  
1            1131.464935  
2             566.472247  
3             529.881344  
4              17.269323  
...                  ...  
4003                 NaN  
4004                 NaN  
4005                 NaN  
4006                 NaN  
4007                 NaN  

[4008 rows x 11 columns]

- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.

In [18]:
# Number of unique values for each column
df.nunique()

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

Columnas categoricas -> Customer, ST, GENDER, Education, Policy Type, Vehicle Class

In [19]:
df["Customer"].unique() # hace referencia al código del cliente y por tanto, hay uno diferente para cada uno

array(['RB50392', 'QZ44356', 'AI49188', ..., 'CW49887', 'MY31220', nan],
      dtype=object)

In [21]:
df["ST"].unique() #Hace referencia al estado en el que vive el cliente y hay 8

array(['Washington', 'Arizona', 'Nevada', 'California', 'Oregon', 'Cali',
       'AZ', 'WA', nan], dtype=object)

In [22]:
df["GENDER"].unique() #Es el genero de la persona, deberia haber unicamente dos. Hay un problema con la notacion

array([nan, 'F', 'M', 'Femal', 'Male', 'female'], dtype=object)

In [24]:
df["Education"].unique() #hay un problema con la notacion tambien, hay repetidas

array(['Master', 'Bachelor', 'High School or Below', 'College',
       'Bachelors', 'Doctor', nan], dtype=object)

In [25]:
df["Policy Type"].unique() #tipo de poliza de seguro, hay 3

array(['Personal Auto', 'Corporate Auto', 'Special Auto', nan],
      dtype=object)

In [26]:
df["Vehicle Class"].unique() #tipo de vehiculo

array(['Four-Door Car', 'Two-Door Car', 'SUV', 'Luxury SUV', 'Sports Car',
       'Luxury Car', nan], dtype=object)

- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.

El estudio descriptivo solo lo podemos hacer de las variables cuantitativas y no las cualitativas. En el resumen, podemos observar que la media de "Income": 39295.701214 y tambien podemos observar que este caso un outlier sera el maximo, pues hay mucha diferencia entre la media con ese valor. Además, de reconocer que el 75% de los usarios esta por debajo de un valor mucho mas pequeño. 

In [30]:
df.describe(include="float")

Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678

- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [31]:
df.describe(include = "object")

Customer      ST GENDER Education Customer Lifetime Value  \
count      1071    1071    954      1071                    1068   
unique     1071       8      5         6                    1027   
top     RB50392  Oregon      F  Bachelor              445811.34%   
freq          1     320    457       324                       4   

       Number of Open Complaints    Policy Type  Vehicle Class  
count                       1071           1071           1071  
unique                         6              3              6  
top                       1/0/00  Personal Auto  Four-Door Car  
freq                         830            780            576

Que la mayoria de clientes soy mujeres, la educación = Bachelor, el número de quejas = 1, poseen coches de cuatro puertas y su poliza es de tipo Personal

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [43]:
less_common_location = pd.Series(df["ST"])
top_5 = less_common_location.value_counts().sort_values(ascending=True)[:5]
top_5

ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [52]:
policy_types = pd.Series(df["Policy Type"]).value_counts().sort_values(ascending=False)
max_number_policies_sold = policy_types[:1]
max_number_policies_sold

Policy Type
Personal Auto    780
Name: count, dtype: int64

### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [53]:
df["Policy Type"].unique()

array(['Personal Auto', 'Corporate Auto', 'Special Auto', nan],
      dtype=object)

In [68]:
personal = df[df["Policy Type"] == "Personal Auto"]
average_personal = personal["Income"].mean()

In [69]:
corporate = df[df["Policy Type"] == "Corporate Auto"]
average_corporate = corporate["Income"].mean()

In [70]:
print(f"Average income  Personal Policy's costumer: {average_personal}")
print(f"Average Income Corporate Policy's costumer: {average_corporate}")

# En media, el salario de los usarios que tiene una poliza corporte es mayor que de aquellos que tienen una personal

Average income  Personal Policy's costumer: 38180.69871794872
Average Income Corporate Policy's costumer: 41390.31196581197


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

>75% - > 532.800000

In [81]:
quantile_75 = df["Total Claim Amount"].quantile(0.75)
quantile_75

532.8

In [72]:
df.describe(include='all')

Customer      ST GENDER Education Customer Lifetime Value  \
count      1071    1071    954      1071                    1068   
unique     1071       8      5         6                    1027   
top     RB50392  Oregon      F  Bachelor              445811.34%   
freq          1     320    457       324                       4   
mean        NaN     NaN    NaN       NaN                     NaN   
std         NaN     NaN    NaN       NaN                     NaN   
min         NaN     NaN    NaN       NaN                     NaN   
25%         NaN     NaN    NaN       NaN                     NaN   
50%         NaN     NaN    NaN       NaN                     NaN   
75%         NaN     NaN    NaN       NaN                     NaN   
max         NaN     NaN    NaN       NaN                     NaN   

              Income  Monthly Premium Auto Number of Open Complaints  \
count    1071.000000           1071.000000                      1071   
unique           NaN                   NaN                         6   
top              NaN                   NaN                    1/0/00   
freq             NaN                   NaN                       830   
mean    39295.701214            193.234360                       NaN   
std     30469.427060           1601.190369                       NaN   
min         0.000000             61.000000                       NaN   
25%     14072.000000             68.000000                       NaN   
50%     36234.000000             83.000000                       NaN   
75%     64631.000000            109.500000                       NaN   
max     99960.000000          35354.000000                       NaN   

          Policy Type  Vehicle Class  Total Claim Amount  
count            1071           1071         1071.000000  
unique              3              6                 NaN  
top     Personal Auto  Four-Door Car                 NaN  
freq              780            576                 NaN  
mean              NaN            NaN          404.986909  
std               NaN            NaN          293.027260  
min               NaN            NaN            0.382107  
25%               NaN            NaN          202.157702  
50%               NaN            NaN          354.729129  
75%               NaN            NaN          532.800000  
max               NaN            NaN         2893.239678

In [82]:
max_claim = df[df["Total Claim Amount"]>quantile_75]
max_claim #aqui estan los clientes que tiene un total Claim Amount superior al 75%

Customer          ST GENDER Education Customer Lifetime Value   Income  \
1     QZ44356     Arizona      F  Bachelor              697953.59%      0.0   
2     AI49188      Nevada      F  Bachelor             1288743.17%  48767.0   
17    OE15005        Cali    NaN   College              394524.16%  28855.0   
23    TZ98966      Nevada    NaN  Bachelor              245019.10%      0.0   
26    US89481  California    NaN  Bachelor              394637.21%      0.0   
...       ...         ...    ...       ...                     ...      ...   
1059  YG44474      Oregon      M   College             1401472.13%  54193.0   
1061  RY92647        Cali      F  Bachelor             1050677.17%      0.0   
1068  GS98873     Arizona      F  Bachelor              323912.47%  16061.0   
1069  CW49887  California      F    Master              462680.11%  79487.0   
1070  MY31220  California      F   College              899704.02%  54230.0   

      Monthly Premium Auto Number of Open Complaints     Policy Type  \
1                     94.0                    1/0/00   Personal Auto   
2                    108.0                    1/0/00   Personal Auto   
17                   101.0                    1/0/00   Personal Auto   
23                    73.0                    1/3/00  Corporate Auto   
26                   111.0                    1/0/00   Personal Auto   
...                    ...                       ...             ...   
1059                 117.0                    1/0/00  Corporate Auto   
1061                  92.0                    1/0/00   Personal Auto   
1068                  88.0                    1/0/00   Personal Auto   
1069                 114.0                    1/0/00    Special Auto   
1070                 112.0                    1/0/00   Personal Auto   

      Vehicle Class  Total Claim Amount  
1     Four-Door Car         1131.464935  
2      Two-Door Car          566.472247  
17              SUV          647.442031  
23    Four-Door Car          554.376763  
26    Four-Door Car          799.200000  
...             ...                 ...  
1059            SUV          720.752945  
1061  Four-Door Car          546.524896  
1068  Four-Door Car          633.600000  
1069            SUV          547.200000  
1070   Two-Door Car          537.600000  

[264 rows x 11 columns]

In [76]:
max_claim.describe(include="all")

Customer      ST GENDER Education Customer Lifetime Value  \
count       264     264    238       264                     264   
unique      264       7      5         5                     256   
top     QZ44356  Oregon      F  Bachelor              578018.22%   
freq          1      90    115        85                       3   
mean        NaN     NaN    NaN       NaN                     NaN   
std         NaN     NaN    NaN       NaN                     NaN   
min         NaN     NaN    NaN       NaN                     NaN   
25%         NaN     NaN    NaN       NaN                     NaN   
50%         NaN     NaN    NaN       NaN                     NaN   
75%         NaN     NaN    NaN       NaN                     NaN   
max         NaN     NaN    NaN       NaN                     NaN   

              Income  Monthly Premium Auto Number of Open Complaints  \
count     264.000000            264.000000                       264   
unique           NaN                   NaN                         6   
top              NaN                   NaN                    1/0/00   
freq             NaN                   NaN                       206   
mean    23677.344697            165.193182                       NaN   
std     27013.483721            623.930992                       NaN   
min         0.000000             63.000000                       NaN   
25%         0.000000             99.000000                       NaN   
50%     18807.000000            114.000000                       NaN   
75%     42423.750000            133.250000                       NaN   
max     99316.000000          10202.000000                       NaN   

          Policy Type Vehicle Class  Total Claim Amount  
count             264           264          264.000000  
unique              3             6                 NaN  
top     Personal Auto           SUV                 NaN  
freq              191           101                 NaN  
mean              NaN           NaN          782.228263  
std               NaN           NaN          292.751640  
min               NaN           NaN          537.600000  
25%               NaN           NaN          606.521741  
50%               NaN           NaN          679.597985  
75%               NaN           NaN          851.400000  
max               NaN           NaN         2893.239678

In [83]:
max_claim.shape

(264, 11)